agent/src/main/java/com/newrelic/agent/android/util/OfflineStorage.java

```diff
@@ -0,0 +1,161 @@
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.util.HashMap;
+import java.util.Map;
+
+public class OfflineStorage {
+    private static final String OFFLINE_STORAGE = "nr_offline_storage";
+    private static final int DEFAULT_MAX_OFFLINE_Storage_SIZE = 100 * 1024 * 1024; //MB
+    private static final AgentLog log = AgentLogManager.getAgentLog();
+    private static File offlineStorage;
+    private static String offlineFilePath = "";
+    private static int offlineStorageSize = 100 * 1024 * 1024; //MB
+
+    public OfflineStorage(Context context) {
+        try {
+            offlineStorage = new File(context.getFilesDir(), OFFLINE_STORAGE);
+            if (!offlineStorage.exists()) {
+                offlineStorage.mkdirs();
+            }
+        } catch (Exception ex) {
+            log.error("OfflineStorage: ", ex);
+        }
+    }
+
+    public boolean persistDataToDisk(String data) {
+        boolean isSaved = false;
+        try {
+            //TODO: Stop collecting data - need to calculate incoming data as well
+            if (getTotalFileSize() > DEFAULT_MAX_OFFLINE_Storage_SIZE) {
+                return false;
+            }
+
+            if (!offlineStorage.exists()) {
+                offlineStorage.mkdirs();
+            }
+
+            File offlineFile = new File(offlineStorage.getAbsolutePath(), "payload_" + System.currentTimeMillis());
+            if (!offlineFile.exists()) {
+                offlineFile.createNewFile();
+                setOfflineFilePath(offlineFile.getAbsolutePath());
+            }
+
+            BufferedWriter buf = new BufferedWriter(new FileWriter(offlineFile, true));
+            buf.write(data);
+            buf.close();
+            isSaved = true;
+        } catch (Exception e) {
+            log.error("OfflineStorage: ", e);
+            isSaved = false;
+        }
+        return isSaved;
+    }
+
+    public Map<String, String> getAllOfflineData() {
+        Map<String, String> harvestDataObjects = new HashMap<String, String>();
+        try {
+            if (offlineStorage == null) {
+                return harvestDataObjects;
+            }
+
+            File[] files = offlineStorage.listFiles();
+            if (files.length > 0) {
+                for (int i = 0; i < files.length; i++) {
+                    BufferedReader in = null;
+                    try {
+                        in = new BufferedReader(new FileReader(files[i]));
+                        String harvestDataFromFile = in.readLine();
+                        harvestDataObjects.put(files[i].getAbsolutePath(), harvestDataFromFile);
+                    } catch (Exception e) {
+                        log.error("OfflineStorage: ", e);
+                    }
+
+                }
+            }
+        } catch (Exception e) {
+            log.error("OfflineStorage: ", e);
+        }
+        return harvestDataObjects;
+    }
+
+    public double getTotalFileSize() {
```

Comment:
@ywang-nr are you checking for handled exceptions and crash files ?

Reply:
Good catch, I will check those too.

Reply:
After talking to Cameron, we have realized that handledexception and crashes should always be saved, it needs to be excluded outside the DEFAULT_MAX_SIZE.